core/state/state_object.go

```diff
@@ -241,7 +246,8 @@ func (self *StateObject) Copy(db trie.Database, onDirty func(addr common.Address
 	stateObject := NewObject(self.address, self.data, onDirty)
 	stateObject.trie = self.trie
 	stateObject.code = self.code
-	stateObject.storage = self.storage.Copy()
+	stateObject.dirtyStorage = self.dirtyStorage.Copy()
+	stateObject.cachedStorage = self.dirtyStorage.Copy()
```

Comment:
Indeed, we only snapshot the changes to restore to, since copying the non-changed values is exactly what we want to avoid. This will entail that rollbacks will flush the read cache.